### DISCUSSION TITLE
Build KwArgs for function call in embedded

### DISCUSSION THREAD

**Discussion by subatiq** [RustPython/RustPython#5241] [Q&A]
I'm stuck on trying to build KwArgs or PyDict to pass to function call. In python module I have

```python
def run(**kwargs) -> None:
    pass
```

`kwargs` are all strings. I try to call it from Rust:

```rust
interpreter.enter(move |vm| {
            let module = vm.import("main", 0).unwrap();
            module.get_attr("run", vm).unwrap().call(config, vm);

        });
```

`config` initially has type `HashMap<String, String>`, though I've tried converting it into `IndexMap` to then build KwArgs from it. The closest I've got is 

```rust 
struct ConfigArg(HashMap<String, String>);

impl Into<KwArgs> for ConfigArg {
    fn into(self) -> KwArgs {
        let mut map = IndexMap::new();

        let mut args = ArgMapping::
        for (key, value) in self.0 {
            let value = PyStr::from(value);
            map.insert(key, value);
        }

        KwArgs::new(map)
    }
}
```

The problem above is that `KwArgs::new(...)` expects `IndexMap<String, PyObjectRef>` and after reading docs I'm still not sure how to achieve this.

**Comment by youknowone** [comment]
For constants, probably FromIterator will be the easiest way to build it. https://docs.rs/rustpython-vm/latest/rustpython_vm/function/struct.KwArgs.html#impl-FromIterator%3C(String,+T)%3E-for-KwArgs%3CT%3E

```rust
let kwargs: KwArgs = [
    ("a".to_owned(), vm.new_pyobj("v1".to_owned()),
    ("b".to_owned(), vm.new_pyobj("v2".to_owned()),
].into_iter().collect();
```

Otherwise, building IndexMap will be also possible. Don't forget to convert String to PyObejctRef.
```
impl Into<KwArgs> for ConfigArg {
    fn into(self) -> KwArgs {
        let mut map = IndexMap::new();
        for (key, value) in self.0 {
            let value = vm.new_pyobj(value);  // or `vm.ctx.new_str(value).into()`
            map.insert(key, value);
        }
        KwArgs::new(map)
    }
}
```
